extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/TableEditor.java

```diff
@@ -0,0 +1,369 @@
+  }
+
+  private long dropColumns(List<String> columnsToDrop) throws CatalogException
+  {
+    if (CollectionUtils.isNullOrEmpty(columnsToDrop)) {
+      return 0;
+    }
+    return catalog.tables().updateColumns(
+        id,
+        table -> applyDropColumns(table, columnsToDrop)
+    );
+  }
+
+  private TableSpec applyDropColumns(final TableMetadata table, final List<String> toDrop)
+  {
+    final TableSpec existingSpec = table.spec();
+    List<ColumnSpec> existingColumns = existingSpec.columns();
+    if (CollectionUtils.isNullOrEmpty(existingColumns)) {
+      return null;
+    }
+    Set<String> drop = new HashSet<String>(toDrop);
+    List<ColumnSpec> revised = new ArrayList<>();
+    for (ColumnSpec col : existingColumns) {
+      if (!drop.contains(col.name())) {
+        revised.add(col);
+      }
+    }
+    if (revised.size() == existingColumns.size()) {
+      // Nothing changed
+      return null;
+    }
+    return existingSpec.withColumns(revised);
+  }
+
+
+  private long updateProperties(Map<String, Object> updates) throws CatalogException
+  {
+    if (updates == null || updates.isEmpty()) {
+      return 0;
+    }
+    return catalog.tables().updateProperties(
+        id,
+        table -> applyUpdateProperties(table, updates)
+    );
+  }
+
+  private TableSpec applyUpdateProperties(
+      final TableMetadata table,
+      final Map<String, Object> updates
+  ) throws CatalogException
+  {
+    final TableSpec existingSpec = table.spec();
+    final TableDefn defn = resolveDefn(existingSpec.type());
+    final Map<String, Object> revised = defn.mergeProperties(
+        existingSpec.properties(),
+        updates
+    );
+    try {
+      defn.validate(revised, catalog.jsonMapper());
+    }
+    catch (IAE e) {
+      throw CatalogException.badRequest(e.getMessage());
+    }
+    return existingSpec.withProperties(revised);
+  }
+
+  private TableDefn resolveDefn(String tableType) throws CatalogException
+  {
+    TableDefn defn = catalog.tableRegistry().defnFor(tableType);
+    if (defn == null) {
+      throw new CatalogException(
+          CatalogException.BAD_STATE,
+          Response.Status.INTERNAL_SERVER_ERROR,
+          "Table %s has an invalid type [%s]",
+          id.sqlName(),
+          tableType
+      );
+    }
+    return defn;
+  }
+
+  private long updateColumns(final List<ColumnSpec> updates) throws CatalogException
+  {
+    if (CollectionUtils.isNullOrEmpty(updates)) {
+      return 0;
+    }
+    return catalog.tables().updateColumns(
+        id,
+        table -> applyUpdateColumns(table, updates)
+    );
+  }
+
+  private TableSpec applyUpdateColumns(
+      final TableMetadata table,
+      final List<ColumnSpec> updates
+  ) throws CatalogException
+  {
+    final TableSpec existingSpec = table.spec();
+    final TableDefn defn = resolveDefn(existingSpec.type());
+    final List<ColumnSpec> revised = defn.mergeColumns(existingSpec.columns(), updates);
```

Comment:
How do we protect a user from dropping a column, and a column with the same name back, after. In this case, if the table is a Druid datasource,  the data for the previous column may not have been deleted, in which cause querying the column will return old data, whereas the user would likely assume the data should be null. Should we document this, or store a set of columns that have been previously deleted?

Reply:
makes sense.